src/Mod/Fem/femmesh/gmshtools.py

```diff
@@ -733,6 +733,13 @@ def write_geo(self):
         geo = open(self.temp_file_geo, "w")
         geo.write("// geo file for meshing with Gmsh meshing software created by FreeCAD\n")
         geo.write("\n")
+
+        cpuCount = os.cpu_count()
```

Comment:
From the docs of `os.cpu_count()`:
> This number is not equivalent to the number of CPUs the current process can use. The number of usable CPUs can be obtained with len(os.sched_getaffinity(0))

Reply:
Locking processes to specific CPU:s may be done on a realtime server, but I doubt that anyone does that on a computer that runs regular applications like CAD software. So I'm not sure it's worthwhile to make the code less readable for a use case that doesn't really exist. In addition multiprocessing.cpu_count() is already used when starting Calculix multi-threaded (which is a wrapper of os.cpu_count())

Reply:
There are people who run these processes under containerized workflows.

Reply:
*  sched_affinity is a unix/posix-specific function, as said in the docs "note: This method is only available on some UNIX platforms."
* cpu_count() is already used in FreeCAD.
* nothing will break if you in some special situation assign more threads than CPUs, it's just somewhat less optimal.

Reply:
Ok, just confirmed that `os.cpu_count()` actually shows correct number of allowed cores in the lxd container.

Reply:
Thanks, great to know, I did not know that.

Reply:
You could still add a comment mentioning where this might cause problems. Additionally, even for people using it on a "computer that does regular operations", we may still want to have some free cores (like we do when compiling with `make -j$(nproc --ignore=2)`). I say `cpuCount=None` can be an argument (also please keep to snake case and use `cpu_count` unless that causes issues) so the user has an option to set it when running.

Reply:
If there is some policy for FreeCAD to not use all cores I think you should implement that top-down with a global setting in preferences. There is no such setting. Until then I think it's much better to use all available cores rather than just one like now. When I run gmsh I want it to finish as quickly as possible.

I don't think it's necessary with a comment. Using all cores for OMP processing is standard procedure. It's not like the computer will go suddenly unresponsive as there's no I/O processing, it's just pure maths calculations i RAM which the OS has no issues scheduling together with user interface tasks.

I'll change the patch to snake case.

Reply:
> If there is some policy for FreeCAD to not use all cores I think you should implement that top-down with a global setting in preferences. There is no such setting. Until then I think it's much better to use all available cores rather than just one like now. When I run gmsh I want it to finish as quickly as possible.

FreeCAD as a program itself is for the most part serial and calls for more cores only at certain times. You are correct that we need some such setting so that it doesn't hog all processing power for a long time. It is also possible to want GMSH/Calculix and other external programs to use more cores, or run on a different computer altogether.

> 
> I don't think it's necessary with a comment. Using all cores for OMP processing is standard procedure. It's not like the computer will go suddenly unresponsive as there's no I/O processing, it's just pure maths calculations i RAM which the OS has no issues scheduling together with user interface tasks.

You would be surprised. "Just pure math calculations" is basically the crux of all scientific computing that can up millions of cores. One only needs to give a mesh size too small and they're stuck with no response for a long time.

> 
> I'll change the patch to snake case.

Thanks.

Reply:
I don't think we should be arguing about how scheduling of threads work. These type of calculations in RAM with no I/O using all available cores will not lead to an unresponsive computer unless you run out of memory, in which case the computer gets unresponsive even with just one thread. This is not at all the same as parallel compilation for example which is a very I/O-intensive process. With I/O-intensive operations multiple threads can indeed bog down the computer. With no I/O, the OS kernel has no trouble scheduling the interactive foreground processes responsively even with 100% load on all cores.

Running the mesher and solver multi-threaded is a big time saver with virtually no drawbacks.

I think that desktop software should be as performant as possible and should not be made artificially slower just for some special use case that very few users will be interested in. And indeed, FreeCAD's default setting for Calculix is to run with all available cores. How many users goes in and changes that number? (Well, actually I did change it to match my core count because I didn't realize that the default value 1 means use all cores rather than just one core, but that's another issue).

I'm not against having multi-core configuration in the preferences for these rare special needs, but it's not good like now have it only for Calculix (and not well-formulated what the setting actually does), so I think one should file a separate bug/issue about multi-core configuration to consolidate it into one. I would myself consider that a low priority feature though, while actually running multi-threaded per default is a high priority and a very easy fix which should be a no-brainer. In my current FEM/modeling project I have run the mesher 100s of times, I save almost 2 minutes each time (2:30 to down to 40 seconds).

Reply:
To add to the discussion, for at least like the past 15 years computers have got increasingly parallel with more cores as it has not been possible to scale up the speed of a single core as much as was possible back in say the 1990s when I started out. This means that to make efficient desktop software you should write the software to use as many cores as possible when it improves performance.

As an example, the open-source raw converter rawtherapee that I contributed to over a few years we had a guy whose only job was to parallelize and optimize existing code to make the algorithms use as many cores as possible as efficiently as possible. It had results and the software became very responsive. No user ever complained about the software using too many cores, but instead rejoiced over the performance improvements.

I have used OMP and other parallelism APIs myself in many software projects to get more responsive software, it's a natural part of modern software development. These days you even look at putting parallel algorithms on the graphics card (GPU) when possible.

If FreeCAD is truly intended to be a single-threaded software for most part and that is seen as a desirable feature rather than something that limits performance, I would be most surprised.